File: frontend/src/api/configurations.ts
import { client } from "./client";
import type { components } from "./schema";
import type { OverlayData } from "./preview";

// ---- Type exports --------------------------------------------------------

export type ConditionMapResponse =
  components["schemas"]["ConditionMapResponse"];
export type ConditionMapCreate =
  components["schemas"]["ConditionMapCreate"];
export type ConditionMapUpdate =
  components["schemas"]["ConditionMapUpdate"];

export type ConditionResponse = components["schemas"]["ConditionResponse"];
export type ConditionCreate = components["schemas"]["ConditionCreate"];
export type ConditionUpdate = components["schemas"]["ConditionUpdate"];

export type CaseResponse = components["schemas"]["CaseResponse"];
export type CaseCreate = components["schemas"]["CaseCreate"];
export type CaseUpdate = components["schemas"]["CaseUpdate"];
export type CaseDuplicateRequest = components["schemas"]["CaseDuplicateRequest"];

export type RunResponse = components["schemas"]["RunResponse"];
export type RunCreate = components["schemas"]["RunCreate"];
export type RunUpdate = components["schemas"]["RunUpdate"];

export type DiffResult = components["schemas"]["DiffResult"];
export type DiffField = components["schemas"]["DiffField"];
export type CaseCompareResult = components["schemas"]["CaseCompareResult"];
export type PartsDiffResult = components["schemas"]["PartsDiffResult"];

export type ConditionMapFolderResponse = components["schemas"]["ConditionMapFolderResponse"];
export type ConditionMapFolderCreate = components["schemas"]["ConditionMapFolderCreate"];
export type ConditionMapFolderUpdate = components["schemas"]["ConditionMapFolderUpdate"];

export type CaseFolderResponse = components["schemas"]["CaseFolderResponse"];
export type CaseFolderCreate = components["schemas"]["CaseFolderCreate"];
export type CaseFolderUpdate = components["schemas"]["CaseFolderUpdate"];

export type SyncRunsPreview = components["schemas"]["SyncRunsPreview"];
export type SyncRunsPreviewItem = components["schemas"]["SyncRunsPreviewItem"];

// ---- ConditionMap API ---------------------------------------------------

export const mapsApi = {
  list: (): Promise<ConditionMapResponse[]> =>
    client.get("/maps/"),

  get: (mapId: string): Promise<ConditionMapResponse> =>
    client.get(`/maps/${mapId}`),

  create: (data: ConditionMapCreate): Promise<ConditionMapResponse> =>
    client.post("/maps/", data),

  update: (mapId: string, data: ConditionMapUpdate): Promise<ConditionMapResponse> =>
    client.patch(`/maps/${mapId}`, data),

  delete: (mapId: string): Promise<void> =>
    client.delete(`/maps/${mapId}`),

  updateFolder: (mapId: string, folderId: string | null): Promise<ConditionMapResponse> =>
    client.patch(`/maps/${mapId}`, { folder_id: folderId }),
};

// ---- ConditionMap Folder API --------------------------------------------

export const mapFoldersApi = {
  list: (): Promise<ConditionMapFolderResponse[]> =>
    client.get("/maps/folders/"),

  create: (data: ConditionMapFolderCreate): Promise<ConditionMapFolderResponse> =>
    client.post("/maps/folders/", data),

  update: (folderId: string, data: ConditionMapFolderUpdate): Promise<ConditionMapFolderResponse> =>
    client.patch(`/maps/folders/${folderId}`, data),

  delete: (folderId: string): Promise<void> =>
    client.delete(`/maps/folders/${folderId}`),
};

// ---- Condition API ------------------------------------------------------

export const conditionsApi = {
  list: (mapId: string): Promise<ConditionResponse[]> =>
    client.get(`/maps/${mapId}/conditions/`),

  get: (mapId: string, conditionId: string): Promise<ConditionResponse> =>
    client.get(`/maps/${mapId}/conditions/${conditionId}`),

  create: (mapId: string, data: ConditionCreate): Promise<ConditionResponse> =>
    client.post(`/maps/${mapId}/conditions/`, data),

  update: (
    mapId: string,
    conditionId: string,
    data: ConditionUpdate,
  ): Promise<ConditionResponse> =>
    client.patch(`/maps/${mapId}/conditions/${conditionId}`, data),

  delete: (mapId: string, conditionId: string): Promise<void> =>
    client.delete(`/maps/${mapId}/conditions/${conditionId}`),
};

// ---- Case API -----------------------------------------------------------

export const casesApi = {
  list: (): Promise<CaseResponse[]> =>
    client.get("/cases/"),

  get: (caseId: string): Promise<CaseResponse> =>
    client.get(`/cases/${caseId}`),

  create: (data: CaseCreate, withRuns = false): Promise<CaseResponse> =>
    client.post(`/cases/?with_runs=${withRuns}`, data),

  update: (caseId: string, data: CaseUpdate): Promise<CaseResponse> =>
    client.patch(`/cases/${caseId}`, data),

  delete: (caseId: string): Promise<void> =>
    client.delete(`/cases/${caseId}`),

  duplicate: (caseId: string, data: CaseDuplicateRequest): Promise<CaseResponse> =>
    client.post(`/cases/${caseId}/duplicate`, data),

  updateFolder: (caseId: string, folderId: string | null): Promise<CaseResponse> =>
    client.patch(`/cases/${caseId}`, { folder_id: folderId }),

  compare: (caseId: string, withCaseId: string): Promise<CaseCompareResult> =>
    client.get(`/cases/${caseId}/compare?with=${withCaseId}`),

  syncPreview: (caseId: string, newMapId: string): Promise<SyncRunsPreview> =>
    client.get(`/cases/${caseId}/sync-preview?new_map_id=${encodeURIComponent(newMapId)}`),
};

// ---- Case Folder API ----------------------------------------------------

export const caseFoldersApi = {
  list: (): Promise<CaseFolderResponse[]> =>
    client.get("/cases/folders/"),

  create: (data: CaseFolderCreate): Promise<CaseFolderResponse> =>
    client.post("/cases/folders/", data),

  update: (folderId: string, data: CaseFolderUpdate): Promise<CaseFolderResponse> =>
    client.patch(`/cases/folders/${folderId}`, data),

  delete: (folderId: string): Promise<void> =>
    client.delete(`/cases/folders/${folderId}`),
};

// ---- Run API ------------------------------------------------------------

export const runsApi = {
  list: (caseId: string): Promise<RunResponse[]> =>
    client.get(`/cases/${caseId}/runs/`),

  create: (caseId: string, data: RunCreate): Promise<RunResponse> =>
    client.post(`/cases/${caseId}/runs/`, data),

  update: (caseId: string, runId: string, data: RunUpdate): Promise<RunResponse> =>
    client.patch(`/cases/${caseId}/runs/${runId}`, data),

  generate: (caseId: string, runId: string, geometryOnly = false): Promise<RunResponse> =>
    client.post(`/cases/${caseId}/runs/${runId}/generate?geometry_only=${geometryOnly}`, {}),

  delete: (caseId: string, runId: string): Promise<void> =>
    client.delete(`/cases/${caseId}/runs/${runId}`),

  reset: (caseId: string, runId: string): Promise<RunResponse> =>
    client.post(`/cases/${caseId}/runs/${runId}/reset`, {}),

  /** Generate 5-belt STL for a Run. Must be called before Transform/XML generation. */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  generateBelts: (caseId: string, runId: string): Promise<any> =>
    client.post(`/cases/${caseId}/runs/${runId}/generate-belts`, {}),

  /** Apply ride-height + yaw transform for a Run. No body needed — backend derives all params. */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  transform: (caseId: string, runId: string): Promise<any> =>
    client.post(`/cases/${caseId}/runs/${runId}/transform`, {}),

  download: async (caseId: string, runId: string): Promise<Blob> => {
    const token = localStorage.getItem("vam_token");
    const res = await fetch(`/api/v1/cases/${caseId}/runs/${runId}/download`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      throw new Error(body.detail || `HTTP ${res.status}`);
    }
    return res.blob();
  },

  downloadStl: async (caseId: string, runId: string): Promise<Blob> => {
    const token = localStorage.getItem("vam_token");
    const res = await fetch(`/api/v1/cases/${caseId}/runs/${runId}/download-stl`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      throw new Error(body.detail || `HTTP ${res.status}`);
    }
    return res.blob();
  },

  diff: (runIdA: string, runIdB: string): Promise<DiffResult> =>
    client.get(`/runs/diff?a=${runIdA}&b=${runIdB}`),

  /** Fetch the axes-GLB for a ready Run. Returns a blob URL (caller must revokeObjectURL). */
  getAxesGlbUrl: async (caseId: string, runId: string): Promise<string> => {
    const token = localStorage.getItem("vam_token");
    const res = await fetch(`/api/v1/cases/${caseId}/runs/${runId}/axes-glb`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      throw new Error(body.detail || `HTTP ${res.status}`);
    }
    return URL.createObjectURL(await res.blob());
  },

  /** Fetch overlay data for a ready Run (parsed from its generated XML). */
  getOverlayData: (caseId: string, runId: string): Promise<OverlayData> =>
    client.get<OverlayData>(`/cases/${caseId}/runs/${runId}/overlay`),

  /** Fetch the belt GLB for a Run that has belt_stl_path set. Returns a blob URL (caller must revokeObjectURL). */
  getBeltGlbUrl: async (caseId: string, runId: string): Promise<string> => {
    const token = localStorage.getItem("vam_token");
    const res = await fetch(`/api/v1/cases/${caseId}/runs/${runId}/belt-glb`, {
      headers: token ? { Authorization: `Bearer ${token}` } : {},
    });
    if (!res.ok) {
      const body = await res.json().catch(() => ({}));
      throw new Error(body.detail || `HTTP ${res.status}`);
    }
    return URL.createObjectURL(await res.blob());
  },
};

File: backend/app/api/v1/configurations.py
"""
API router: ConditionMap / Condition / Case / Run management
"""
from fastapi import APIRouter, BackgroundTasks, Depends, Query
from fastapi.responses import FileResponse, Response
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.auth.deps import get_current_user
from app.database import get_db
from app.models.configuration import Case, ConditionMap, Run
from app.models.user import User
from app.schemas.configuration import (
    CaseCreate, CaseResponse, CaseUpdate, CaseDuplicateRequest,
    CaseFolderCreate, CaseFolderUpdate, CaseFolderResponse,
    CaseCompareResult,
    ConditionCreate, ConditionResponse, ConditionUpdate,
    ConditionMapCreate, ConditionMapResponse, ConditionMapUpdate,
    ConditionMapFolderCreate, ConditionMapFolderUpdate, ConditionMapFolderResponse,
    DiffResult,
    RunCreate, RunResponse, RunUpdate,
    SyncRunsPreview,
)
from app.services import configuration_service

router = APIRouter()


# ---------------------------------------------------------------------------
# ConditionMap endpoints  /maps/...
# ---------------------------------------------------------------------------

@router.get("/maps/", response_model=list[ConditionMapResponse])
def list_maps(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    maps = configuration_service.list_maps(db)
    result = []
    for m in maps:
        data = ConditionMapResponse.model_validate(m)
        data.condition_count = len(m.conditions)
        result.append(data)
    return result


@router.post("/maps/", response_model=ConditionMapResponse, status_code=201)
def create_map(
    data: ConditionMapCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    m = configuration_service.create_map(db, data, current_user)
    out = ConditionMapResponse.model_validate(m)
    out.condition_count = len(m.conditions)
    return out



# ---------------------------------------------------------------------------
# Map Folders — must be declared BEFORE /maps/{map_id}
# ---------------------------------------------------------------------------

@router.get("/maps/folders/", response_model=list[ConditionMapFolderResponse])
def list_map_folders(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return configuration_service.list_map_folders(db)


@router.post("/maps/folders/", response_model=ConditionMapFolderResponse, status_code=201)
def create_map_folder(
    data: ConditionMapFolderCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.create_map_folder(db, data, current_user)


@router.patch("/maps/folders/{folder_id}", response_model=ConditionMapFolderResponse)
def update_map_folder(
    folder_id: str,
    data: ConditionMapFolderUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.update_map_folder(db, folder_id, data, current_user)


@router.delete("/maps/folders/{folder_id}", status_code=204)
def delete_map_folder(
    folder_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    configuration_service.delete_map_folder(db, folder_id, current_user)


@router.get("/maps/{map_id}", response_model=ConditionMapResponse)
def get_map(
    map_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    m = configuration_service.get_map(db, map_id)
    out = ConditionMapResponse.model_validate(m)
    out.condition_count = len(m.conditions)
    return out


@router.patch("/maps/{map_id}", response_model=ConditionMapResponse)
def update_map(
    map_id: str,
    data: ConditionMapUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    m = configuration_service.update_map(db, map_id, data, current_user)
    out = ConditionMapResponse.model_validate(m)
    out.condition_count = len(m.conditions)
    return out


@router.delete("/maps/{map_id}", status_code=204)
def delete_map(
    map_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    configuration_service.delete_map(db, map_id, current_user)


# ---------------------------------------------------------------------------
# Condition endpoints  /maps/{map_id}/conditions/...
# ---------------------------------------------------------------------------

@router.get("/maps/{map_id}/conditions/", response_model=list[ConditionResponse])
def list_conditions(
    map_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.list_conditions(db, map_id)


@router.post("/maps/{map_id}/conditions/", response_model=ConditionResponse, status_code=201)
def create_condition(
    map_id: str,
    data: ConditionCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.create_condition(db, map_id, data, current_user)


@router.get("/maps/{map_id}/conditions/{condition_id}", response_model=ConditionResponse)
def get_condition(
    map_id: str,
    condition_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.get_condition(db, map_id, condition_id)


@router.patch("/maps/{map_id}/conditions/{condition_id}", response_model=ConditionResponse)
def update_condition(
    map_id: str,
    condition_id: str,
    data: ConditionUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.update_condition(db, map_id, condition_id, data, current_user)


@router.delete("/maps/{map_id}/conditions/{condition_id}", status_code=204)
def delete_condition(
    map_id: str,
    condition_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    configuration_service.delete_condition(db, map_id, condition_id, current_user)


# ---------------------------------------------------------------------------
# Case endpoints  /cases/...
# ---------------------------------------------------------------------------

@router.get("/cases/", response_model=list[CaseResponse])
def list_cases(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    cases = configuration_service.list_cases(db)
    return [configuration_service.enrich_case_response(db, c) for c in cases]


@router.post("/cases/", response_model=CaseResponse, status_code=201)
def create_case(
    data: CaseCreate,
    with_runs: bool = Query(False, description="Auto-create Runs for all Conditions in the map"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if with_runs:
        case = configuration_service.create_case_with_runs(db, data, current_user)
    else:
        case = configuration_service.create_case(db, data, current_user)
    return configuration_service.enrich_case_response(db, case)



# ---------------------------------------------------------------------------
# Case Folders -- must be declared BEFORE /cases/{case_id}
# ---------------------------------------------------------------------------

@router.get("/cases/folders/", response_model=list[CaseFolderResponse])
def list_case_folders(
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return configuration_service.list_case_folders(db)


@router.post("/cases/folders/", response_model=CaseFolderResponse, status_code=201)
def create_case_folder(
    data: CaseFolderCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.create_case_folder(db, data, current_user)


@router.patch("/cases/folders/{folder_id}", response_model=CaseFolderResponse)
def update_case_folder(
    folder_id: str,
    data: CaseFolderUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.update_case_folder(db, folder_id, data, current_user)


@router.delete("/cases/folders/{folder_id}", status_code=204)
def delete_case_folder(
    folder_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    configuration_service.delete_case_folder(db, folder_id, current_user)


@router.get("/cases/{case_id}", response_model=CaseResponse)
def get_case(
    case_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    c = configuration_service.get_case(db, case_id)
    return configuration_service.enrich_case_response(db, c)


@router.get("/cases/{case_id}/sync-preview", response_model=SyncRunsPreview)
def sync_preview(
    case_id: str,
    new_map_id: str = Query(..., description="ID of the target ConditionMap"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Preview what happens to Runs when this Case switches to a different map."""
    return configuration_service.compute_sync_preview(db, case_id, new_map_id)


@router.patch("/cases/{case_id}", response_model=CaseResponse)
def update_case(
    case_id: str,
    data: CaseUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    c = configuration_service.update_case(db, case_id, data, current_user)
    return configuration_service.enrich_case_response(db, c)


@router.delete("/cases/{case_id}", status_code=204)
def delete_case(
    case_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    configuration_service.delete_case(db, case_id, current_user)


@router.post("/cases/{case_id}/duplicate", response_model=CaseResponse, status_code=201)
def duplicate_case(
    case_id: str,
    data: CaseDuplicateRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    c = configuration_service.duplicate_case(db, case_id, data, current_user)
    return configuration_service.enrich_case_response(db, c)


# ---------------------------------------------------------------------------
# Run endpoints  /cases/{case_id}/runs/...
# ---------------------------------------------------------------------------

@router.get("/cases/{case_id}/runs/", response_model=list[RunResponse])
def list_runs(
    case_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    runs = configuration_service.list_runs(db, case_id)
    return [configuration_service.enrich_run_response(db, r) for r in runs]


@router.post("/cases/{case_id}/runs/", response_model=RunResponse, status_code=201)
def create_run(
    case_id: str,
    data: RunCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    run = configuration_service.create_run(db, case_id, data, current_user)
    return configuration_service.enrich_run_response(db, run)


@router.patch("/cases/{case_id}/runs/{run_id}", response_model=RunResponse)
def update_run(
    case_id: str,
    run_id: str,
    data: RunUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    run = configuration_service.update_run(db, case_id, run_id, data, current_user)
    return configuration_service.enrich_run_response(db, run)


@router.post("/cases/{case_id}/runs/{run_id}/generate", response_model=RunResponse)
def generate_xml(
    case_id: str,
    run_id: str,
    background_tasks: BackgroundTasks,
    geometry_only: bool = Query(False, description="Swap STL only, reuse parent Run XML (requires parent_case_id)"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    run = configuration_service.trigger_xml_generation(
        db, case_id, run_id, current_user, background_tasks, geometry_only
    )
    return configuration_service.enrich_run_response(db, run)


@router.delete("/cases/{case_id}/runs/{run_id}", status_code=204)
def delete_run(
    case_id: str,
    run_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    configuration_service.delete_run(db, case_id, run_id, current_user)


@router.post("/cases/{case_id}/runs/{run_id}/reset", response_model=RunResponse)
def reset_run(
    case_id: str,
    run_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    run = configuration_service.reset_run(db, case_id, run_id, current_user)
    return configuration_service.enrich_run_response(db, run)


@router.get("/cases/{case_id}/runs/{run_id}/download")
def download_xml(
    case_id: str,
    run_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    path = configuration_service.get_xml_path(db, case_id, run_id)
    return FileResponse(
        str(path),
        media_type="application/xml",
        filename=path.name,
    )


@router.get("/cases/{case_id}/runs/{run_id}/download-stl")
def download_stl(
    case_id: str,
    run_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Download the input STL that was used for XML generation."""
    from pathlib import Path as _Path
    from fastapi import HTTPException
    run = db.query(Run).filter_by(id=run_id, case_id=case_id).first()
    if not run:
        raise HTTPException(status_code=404, detail="Run not found")
    if not run.stl_path or not _Path(run.stl_path).exists():
        raise HTTPException(status_code=404, detail="STL file not available for this run")
    p = _Path(run.stl_path)
    return FileResponse(str(p), media_type="application/octet-stream", filename=p.name)


@router.get("/cases/{case_id}/runs/{run_id}/axes-glb")
def get_axes_glb(
    case_id: str,
    run_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return a GLB file containing wheel-rotation-axis arrows and porous-flow-axis arrows."""
    glb_bytes = configuration_service.get_axes_glb(db, case_id, run_id)
    return Response(content=glb_bytes, media_type="model/gltf-binary")


@router.get("/cases/{case_id}/runs/{run_id}/overlay")
def get_run_overlay(
    case_id: str,
    run_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return OverlayData for a ready Run — parsed from its generated XML."""
    return configuration_service.get_run_overlay(db, case_id, run_id)


@router.get("/cases/{case_id}/compare", response_model=CaseCompareResult)
def compare_cases(
    case_id: str,
    with_case: str = Query(..., alias="with", description="ID of the case to compare against"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Compare two cases: template settings diff, map conditions diff, assembly parts diff."""
    return configuration_service.compare_cases(db, case_id, with_case)


# ---------------------------------------------------------------------------
# Diff endpoint
# ---------------------------------------------------------------------------

@router.get("/runs/diff", response_model=DiffResult)
def diff_runs(
    a: str = Query(..., description="Run A ID"),
    b: str = Query(..., description="Run B ID"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return configuration_service.get_diff(db, a, b)
